zip-0321.rst

```diff
@@ -6,5 +6,218 @@
+============
+
+The format must be a valid URI.
+
+The format must permit the representation of one or more (payment address, amount, 
+memo) tuples.
+
+Specification
+=============
+
+The following syntax specification uses ABNF [#RFC5234]_.
+
+URI Syntax:
+
+.. code-block:: EBNF
+
+  zcashurn        = "zcash:" ( zcashaddress [ "?" zcashparams ] / "?" zcashparams )
+  zcashaddress    = 1*base58
+  zcashparams     = zcashparam [ "&" zcashparams ]
+  zcashparam      = [ addrparam / amountparam / memoparam / messageparam / labelparam / reqparam / otherparam ]
+  NONZERO         = %x31-39
+  DIGIT           = %x30-39
+  paramindex      = "." NONZERO *DIGIT
+  addrparam       = "address" [ paramindex ] "=" zcashaddress
+  amountparam     = "amount"  [ paramindex ] "=" 1*DIGIT [ "." 1*8DIGIT ]
+  labelparam      = "label"   [ paramindex ] "=" *qchar
+  memoparam       = "memo"    [ paramindex ] "=" *base64url
+  messageparam    = "message" [ paramindex ] "=" *qchar
+  paramname       = ALPHA *( ALPHA / DIGIT / "+" / "-" )
+  reqparam        = "req-" paramname [ paramindex ] [ "=" *qchar ]
+  otherparam      = paramname [ paramindex ] [ "=" *qchar ]
+  qchar           = unreserved / pct-encoded / allowed-delims / ":" / "@"
+  allowed-delims  = "!" / "$" / "'" / "(" / ")" / "*" / "+" / "," / ";"
+
+Here, ``ALPHA``, ``unreserved`` and ``pct-encoded`` are as defined in
+[#RFC3986]_. "base58" is defined as in [#base58check]_. "base64url" is defined
+as in [#base64url]_ with padding omitted. (Note that this uses a different
+alphabet to the usual base64; the values 62 and 63 in the alphabet are encoded
+as ``-`` and ``_`` respectively.)
+   
+A ZIP-321 URI represents a request for the construction of a transaction having
+one or more *payments*. In the case that only a single payment is being
+requested, the recipient address MAY be included in the ``hier-part`` component
+of the RFC 3986 URI; otherwise, multiple recipient addresses can be specified
+using ``addrparam`` parameters with different indices. 
+
+Addresses, amounts, labels, and messages sharing the same ``paramindex`` are
+interpreted to be associated with the same payment for the purposes of payment
+construction. Parameters with an empty ``paramindex`` are associated with a
+payment to the address with empty ``paramindex``.
+
+A URI of the form ``zcash:<address>?...`` MUST be considered equivalent to a
+URI of the form ``zcash:?address=<address>&...`` where ``<address>`` is an
+instance of ``zcashaddress``.
+
+If there are any non-address parameters having a given ``paramindex``, then 
+the URI MUST contain an address parameter having that ``paramindex``.
+
+Productions of the form ``1*x`` indicate one or more successive instances of the
+production ``x``. Productions of the form ``<n>*<m>x`` indicate at least `<n>` and
+at most `<m>` instances of production ``x``.
+
+Transfer amount
+---------------
+
+If an amount is provided, it MUST be specified in decimal ZEC. Amounts MUST NOT
```

Comment:
Wouldn't it be better to express the formatting in terms of Grouping/Thousand and decimal separators?

"The whole (integer) part must not contain grouping/thousands separators" or similar. People implementing this will find these terms in the number formatting classes of their language of choice.

Reply:
The `MUST NOT contain commas` bit here is actually redundant with the grammar, but the restriction is not limited to the whole part of the amount. Maybe "Amounts MUST NOT contain commas or other separators apart from the decimal point."?

Reply:
it seems pretty clear to me, but I googled it and it results that some locales use spaces as thousand separators 😱